File: tetris.py
import numpy as np
import random
import copy
from multiprocessing import Pool, cpu_count
import matplotlib.pyplot as plt
import timeit

class Piece():
    def __init__(self, skirt, heights):
        self.skirt = np.array(skirt)
        self.heights = np.array(heights)
        self.width = self.heights.size

class Board():
    def __init__(self):
        self.width = 10
        self.height = 20

        self.state = np.zeros((self.height, self.width), dtype=int)
        self.tops = np.zeros(self.width, dtype=int)
        self.maxHeight = max(self.tops)
        self.linesCleared = 0
        self.totalLinesCleared = 0
        self.fitness = 0

        self.weights = None
        self.features = np.zeros(7)
        self.score = 0

        self.setPieces()
        self.nextPiece = self.pieces[random.randint(0, len(self.pieces) - 1)]
        self.nextNextPiece = self.pieces[random.randint(0, len(self.pieces) - 1)]

    def printState(self):
        for i in self.state[::-1]:
            for j in i:
                print(j, end=' ')
            print()
        print()

    def drop(self, piece, xOffset):
        # returns true if succesful drop
        temp = []
        for i in range(piece.width):
            temp.append(self.tops[xOffset + i] - piece.skirt[i])

        tempMax = max(temp)
        for i in range(len(temp)):
            temp[i] = tempMax - temp[i]

        for i in range(len(temp)):
            temp[i] += self.tops[xOffset + i]

        for j in range(piece.width):
            for i in range(piece.heights[j]):
                if i + temp[j] >= self.height:
                    return False
                self.state[i + temp[j]][j + xOffset] = 1
            self.tops[xOffset + j] = temp[j] + piece.heights[j]

        # update maxHeight
        self.maxHeight = max(self.tops)

        # check for cleared lines
        self.clearLines()

        # update next piece
        self.nextPiece = self.nextNextPiece
        self.nextNextPiece = self.pieces[random.randint(0, len(self.pieces) - 1)]

        return True

    def clearLines(self):
        # checks for full lines, returns number of lines cleared
        count = 0
        i = 0

        while i < self.maxHeight:
            # check if lines are full
            if 0 not in self.state[i]:
                # move lines down
                self.state[i:-1] = self.state[i + 1:]
                self.state[-1] = np.zeros(self.width)
                # keep track of lines cleared
                count += 1
                # update other states
                self.maxHeight -= 1
                self.tops -= 1
            else:
                i += 1
        self.totalLinesCleared += count
        self.linesCleared = count

        scores = {0: 0, 1: 100, 2: 300, 3: 500, 4: 2400}

        self.fitness += scores[count]

    def setPieces(self):
        iPiece1 = Piece([0], [4])
        iPiece2 = Piece([0, 0, 0, 0], [1, 1, 1, 1])
        jPiece1 = Piece([0, 0, 0], [2, 1, 1])
        jPiece2 = Piece([0, 2], [3, 1])
        jPiece3 = Piece([1, 1, 0], [1, 1, 2])
        jPiece4 = Piece([0, 0], [1, 3])
        lPiece1 = Piece([0, 0, 0], [1, 1, 2])
        lPiece2 = Piece([0, 0], [3, 1])
        lPiece3 = Piece([0, 1, 1], [2, 1, 1])
        lPiece4 = Piece([2, 0], [1, 3])
        oPiece1 = Piece([0, 0], [2, 2])
        sPiece1 = Piece([0, 0, 1], [1, 2, 1])
        sPiece2 = Piece([1, 0], [2, 2])
        tPiece1 = Piece([0, 0, 0], [1, 2, 1])
        tPiece2 = Piece([0, 1], [3, 1])
        tPiece3 = Piece([1, 0, 1], [1, 2, 1])
        tPiece4 = Piece([1, 0], [1, 3])
        zPiece1 = Piece([1, 0, 0], [1, 2, 1])
        zPiece2 = Piece([0, 1], [2, 2])

        iPieces = [iPiece1, iPiece2]
        jPieces = [jPiece1, jPiece2, jPiece3, jPiece4]
        lPieces = [lPiece1, lPiece2, lPiece3, lPiece4]
        oPieces = [oPiece1]
        sPieces = [sPiece1, sPiece2]
        tPieces = [tPiece1, tPiece2, tPiece3, tPiece4]
        zPieces = [zPiece1, zPiece2]

        # self.pieces -> piece -> orientation of piece
        self.pieces = [iPieces, jPieces, lPieces, oPieces, sPieces, tPieces, zPieces]

    def calcFeatures(self):
        # calculates the features:

        # number of holes and depth of wholes
        holes = 0
        holeDepths = 0
        # iterate through each col
        for i in range(self.width):
            curr = self.state[:self.tops[i], i]
            # find the zeros
            zeros = np.where(curr == 0)[0]
            # add the number of zeros
            holes += zeros.size
            # add the depth of the zeros, calculated by the self.tops[i] - index
            holeDepths += np.sum(self.tops[i] - zeros)

        # number of blocks(non zero) in column 0
        col0 = np.count_nonzero(self.state[:self.tops[0], 0])

        # bumpiness of the tops, sum and abs
        diff = np.diff(self.tops)
        bumps = np.sum(np.abs(diff))

        # max height
        maxHeight = self.maxHeight

        # lines cleared from this drop
        linesCleared = self.linesCleared

        # if tetris
        tetris = 1 if linesCleared == 4 else 0

        self.features =  np.transpose([[holes, holeDepths, col0, bumps, maxHeight, linesCleared, tetris]])

    def calcScore(self):
        # calculates the score based off the weights and features
        self.score = np.dot(self.weights, self.features)

    def ReLU(self, Z):
        return np.clip(Z, 0, None)

    def calcScoreNN(self):

        W1, b1, W2, b2 = self.weights

        Z1 = W1 @ self.features + b1
        Y1 = self.ReLU(Z1)
        Z2 = W2 @ Y1 + b2
        Y2 = self.ReLU(Z2)

        self.score =  Y2

    def nextState(self):
        # finds the best next state and updates itself to it
        # returns False if game is over
        bestScore = -np.inf
        # select a random piece
        temp = random.randint(0, len(self.pieces) - 1)
        nextPiece = self.pieces[temp]
        # print(temp)

        alive = False

        for orientation in nextPiece:
            for offset in range(self.width - orientation.width + 1):
                # for each possible position, place and append to states
                nextBoard = copy.deepcopy(self)
                if not nextBoard.drop(orientation, offset):
                    continue
                alive = True
                nextBoard.calcFeatures()
                nextBoard.calcScoreNN()
                # nextBoard.calcScore()
                if nextBoard.score > bestScore:
                    bestScore = nextBoard.score
                    bestState = nextBoard

        if alive:
            self.__dict__ = copy.deepcopy(bestState.__dict__)
        return alive

    def nextStateLookAhead(self):
        # finds the best next state by looking at next 2 states and updates itself to next state
        # returns False if no more possible moves

        bestScore, bestState = self.nextStateHelper(1)

        # if we we are still alive
        if bestState is not None:
            self.__dict__ = copy.deepcopy(bestState.__dict__)
            return True
        else:
            return False


    def nextStateHelper(self, depth):
        # recursive helper function for next state
        # returns the next best score and best state (-inf if no possible moves)
        bestScore = -np.inf
        bestState = None
        # bestChild = None

        for orientation in self.nextPiece:
            for offset in range(self.width - orientation.width + 1):
                # for each orientation and position, drop piece
                nextBoard = copy.deepcopy(self)
                # if the drop is not succesful, continue to next
                if not nextBoard.drop(orientation, offset):
                    continue

                # if leaf node
                if depth == 2:
                    nextBoard.calcFeatures()
                    nextBoard.calcScoreNN()
                    # nextBoard.calcScore()
                    if nextBoard.score > bestScore:
                        bestScore = nextBoard.score
                        bestState = nextBoard

                # elif parent, recursievley call on child
                elif depth == 1:
                    childScore, childState = nextBoard.nextStateHelper(2)
                    # if the child is best, than keep track of the nextBoard
                    if childScore > bestScore:
                        bestState = copy.copy(nextBoard)
                        bestScore = childScore
                        # bestChild = childState

        # if depth == 1:
        #     bestState.printState()
        #     bestChild.printState()

        return bestScore, bestState

class GeneticAlgorithm():
    def __init__(self, children=16, mutationRate=0.2, layerDims=[7, 5, 1]):
        self.mutationRate = mutationRate
        # self.childrenWeights = np.random.uniform(-1, 1, (children, 7))

        # randomly generate weights for each child
        self.childrenWeights = []
        for child in range(children):
            childWeights = []
            childWeights.append(np.random.normal(0, 0.01, [layerDims[1], layerDims[0]]))
            childWeights.append(np.zeros((layerDims[1], 1)))
            childWeights.append(np.random.normal(0, 0.01, [layerDims[2], layerDims[1]]))
            childWeights.append(np.zeros((layerDims[2], 1)))
            self.childrenWeights.append(childWeights)

        self.children = int(children)
        self.generations = 0
        self.scores = []
        self.bestScore = 0

    def mutate(self):
        # generate new weights by mutation

        # mutate each parent into n/4 children
        newChildrenWeights = []
        for currWeights in self.parentWeights:
            newChildrenWeights.append(currWeights)
            for i in range(int(len(self.childrenWeights) / 4) - 1):
                newWeights = currWeights + np.random.normal(0, self.mutationRate, 7)
                newWeights = np.clip(newWeights, -1, 1)
                newChildrenWeights.append(newWeights)

        # remainig will be random new weights
        randomWeights = np.random.uniform(-1, 1, (self.children - len(newChildrenWeights), 7))
        newChildrenWeights = np.concatenate((newChildrenWeights, randomWeights))

        self.childrenWeights = newChildrenWeights

    def mutateNN(self):
        newChildrenWeights = []
        for parent in self.parentWeights:
            newChildrenWeights.append(parent)
            for i in range(int(len(self.childrenWeights) / 2) - 1):
                newWeights = []
                for param in parent:
                    mutation =  np.random.normal(0, 0.2, param.shape)
                    newWeights.append(np.clip(param + mutation, -1, 1))
                newChildrenWeights.append(newWeights)

        self.childrenWeights = newChildrenWeights

    def play(self):
        self.generations += 1
        fitnessWeights = []
        with Pool(cpu_count()) as p:
            fitnessWeights.append(p.map(self.playHelper, self.childrenWeights))

        # take top 2 fitness and store their weight
        sortedWeights = sorted(fitnessWeights[0], key=lambda x: x[0], reverse=True)
        self.parentWeights = []
        self.parentWeights.append(sortedWeights[0][1])
        self.parentWeights.append(sortedWeights[1][1])

        self.scores.append(sortedWeights[0][0])

        if sortedWeights[0][0] > self.bestScore:
            self.bestScore = sortedWeights[0][0]
            self.bestWeights = sortedWeights[0][1]

    def playHelper(self, weights):
        # returns [combined fitness, and weights]
        # runs twice to smooth
        totalFitness = 0
        for i in range(2):
            board = Board()
            board.weights = weights

            while board.nextStateLookAhead():
                pass

            totalFitness += board.fitness

        return [totalFitness / 2, board.weights]

    def resume(self):

        with open('record.npy', 'rb') as f:
            self.scores = np.load(f).tolist()
            prev = []
            for i in range(4):
                prev.append(np.load(f))

        print(self.scores)
        print(prev)

        # change from list of lists, to list of np.arrays
        temp = []
        for i in prev:
            temp.append(np.array(i))

        newChildrenWeights = [temp]

        for i in range(int(len(self.childrenWeights)) - 1):
            newWeights = []
            for param in temp:
                mutation =  np.random.normal(0, 0.2, param.shape)
                newWeights.append(np.clip(param + mutation, -1, 1))
            newChildrenWeights.append(newWeights)

        self.childrenWeights = newChildrenWeights

    def output(self):
        with open('record.npy', 'wb') as f:
            np.save(f, self.scores)
            for i in self.parentWeights[0]:
                np.save(f, i)


if __name__ == '__main__':
    ga = GeneticAlgorithm(children=128)

    # ga.resume()
    try:
        ga.resume()
    except:
        print("no record found")

    epochs = 100

    start = timeit.default_timer()

    for i in range(epochs):
        ga.play()
        ga.mutateNN()
        print("epoch: ", i)
        if i % 10 == 0:
            ga.output()
        # ga.mutate()

    print("time: ", timeit.default_timer() - start)
    # print("best score: ", ga.bestScore)
    # print("weights: ", ga.bestWeights)
    print("most recent weights: ", ga.parentWeights[0])

    ga.output()


    plt.plot(ga.scores)
    plt.show()

File: tetris.cpp
#include "tetris.h"
#include <algorithm>
#include <array>
#include <vector>
#include <time.h>
#include <ctime>
#include <random>

Tetris::Tetris()
{
    setPieces();
    seed = std::time(0);
    std::mt19937 random(seed);
    rg = random;
    nextPiece = random() % 7;
    nextNextPiece = random() % 7;

    w1 = new double[5][7]();
    b1 = new double[5]();
    w2 = new double[5]();
    b2 = new double[1]();

    clone = false;
}

Tetris::~Tetris()
{
    if (!clone)
    {
        delete[] w1;
        delete[] b1;
        delete[] w2;
        delete[] b2;
    }
}

Tetris::Tetris(const Tetris &rhs)
{
    width = rhs.width;
    height = rhs.height;
    state = rhs.state;
    tops = rhs.tops;
    maxHeight = rhs.maxHeight;
    hold = rhs.hold;
    justHeld = rhs.justHeld;

    linesCleared = rhs.linesCleared;
    totalLinesCleared = rhs.totalLinesCleared;
    igScore = rhs.igScore;
    tetris = rhs.tetris;
    weights = rhs.weights;
    allPieces = rhs.allPieces;
    nextPiece = rhs.nextPiece;
    nextNextPiece = rhs.nextNextPiece;
    rg = rhs.rg;
    seed = rhs.seed;
    clone = true;

    w1 = rhs.w1;
    b1 = rhs.b1;
    w2 = rhs.w2;
    b2 = rhs.b2;
}

Tetris &Tetris::operator=(const Tetris &rhs)
{
    width = rhs.width;
    height = rhs.height;
    state = rhs.state;
    tops = rhs.tops;
    maxHeight = rhs.maxHeight;
    hold = rhs.hold;
    justHeld = rhs.justHeld;

    linesCleared = rhs.linesCleared;
    totalLinesCleared = rhs.totalLinesCleared;
    igScore = rhs.igScore;
    tetris = rhs.tetris;
    weights = rhs.weights;
    allPieces = rhs.allPieces;
    nextPiece = rhs.nextPiece;
    nextNextPiece = rhs.nextNextPiece;
    rg = rhs.rg;
    seed = rhs.seed;
    clone = true;

    w1 = rhs.w1;
    b1 = rhs.b1;
    w2 = rhs.w2;
    b2 = rhs.b2;
    return *this;
}

void Tetris::printState()
{
    // create bitset
    std::vector<std::bitset<32>> bits(10);
    std::bitset<32> curr;
    for (int i = 0; i < width; ++i)
    {
        curr = state[i];
        bits[i] = curr;
    }

    for (int i = height; i >= 0; --i)
    {
        for (int j = 0; j < width; ++j)
        {
            std::cout << bits[j][i] << " ";
        }
        std::cout << "\n";
    }
    std::cout << "\n";
}

bool Tetris::drop(int pieceNumber, int orientation, int xOffset)
{
    Piece *piece = &allPieces[pieceNumber][orientation];

    // returns true if successful drop
    int temp[4];
    int yOffset = 0;

    for (int i = 0; i < piece->width; ++i)
    {
        temp[i] = tops[xOffset + i] - piece->skirt[i];
        yOffset = std::max(yOffset, temp[i]);
    }

    int block = 0;
    for (int i = 0; i < piece->width; ++i)
    {
        // bitshift left, bottom is lsb
        block = piece->blocks[i];
        block = block << (yOffset + piece->skirt[i]);

        // use bitwise OR to place block
        state[xOffset + i] = state[xOffset + i] | block;

        // return false if too large
        if (state[xOffset + i] > 1048576)
        {
            return false;
        }

        // update tops
        tops[xOffset + i] = 32 - std::countl_zero(state[xOffset + i]);
    }

    // update max height
    maxHeight = *std::max_element(tops.begin(), tops.end());

    // check for cleared lines
    clearLines();

    return true;
}

void Tetris::clearLines()
{
    int count = 0;
    unsigned bitMask = 1;
    unsigned curr = 0;
    bool full = true;
    unsigned save;

    for (int row = 0; row < maxHeight; ++row)
    {
        // std::cout << "row: " << row << " ";
        full = true;
        for (int col = 0; col < width; ++col)
        {
            // if the row at the colu is a 0, break to next row
            if (((state[col] & (bitMask << row)) == 0))
            {
                full = false;
                break;
            }
        }

        if (full)
        {
            // moves rows down, each column at a time
            for (int col = 0; col < width; ++col)
            {
                // save rows below
                save = state[col] & ((bitMask << row) - 1);

                // remove the below roes
                state[col] = (state[col] >> (row + 1)) << (row);

                // restore save
                state[col] = state[col] | save;
            }

            // decrement row to revisit row;
            row--;

            // update maxHeight, count, and tops
            maxHeight--;
            count++;
            for (int i = 0; i < width; ++i)
            {
                tops[i] = 32 - std::countl_zero(state[i]);
            }
        }
    }

    totalLinesCleared += count;
    linesCleared = count;

    if (count == 4)
    {
        tetris = 1;
    }
    else
    {
        tetris = 0;
    }

    if (count > 0)
    {
        combo++;
    }
    else
    {
        combo = 0;
    }

    int scores[5] = {0, 1, 3, 5, 16};

    igScore += (scores[count]);
}

void Tetris::setWeights(std::vector<double> &newWeights)
{
    weights = newWeights;
}

void Tetris::setWeightsNN(std::vector<std::vector<double>> &w1_, std::vector<double> &b1_, std::vector<double> &w2_, std::vector<double> &b2_)
{
    for (int i = 0; i < 5; ++i)
    {
        for (int j = 0; j < 7; ++j)
        {
            w1[i][j] = w1_[i][j];
        }
        b1[i] = b1_[i];
        w2[i] = w2_[i];
    }
    b2[0] = b2_[0];
}

void Tetris::printWeights()
{
    for (int i = 0; i < 5; ++i)
    {
        for (int j = 0; j < 7; ++j)
        {
            std::cout << w1[i][j] << "\t";
        }
        std::cout << "\n";
    }

    for (int i = 0; i < 5; ++i)
    {
        std::cout << b1[i] << "\t";
    }
    std::cout << "\n";

    for (int i = 0; i < 5; ++i)
    {
        std::cout << w2[i] << "\t";
    }

    std::cout << "\n"
              << b2[0] << std::endl;
}

int Tetris::getIgScore()
{
    return igScore;
}

int Tetris::getTetris()
{
    return tetris;
}

double Tetris::calculateLinScore()
{
    // # of holes
    int holes = 0;
    for (int col = 0; col < width; ++col)
    {
        holes += tops[col] - std::popcount(state[col]);
    }

    // depth of holes
    int depth = 0;
    int curr;
    int ones;
    for (int col = 0; col < width; ++col)
    {
        curr = state[col];
        ones = std::popcount(state[col]);
        for (int row = 0; row < maxHeight; ++row)
        {
            if ((curr & 1) == 0)
            {
                depth += ones;
            }
            else
            {
                ones--;
            }
            curr >>= 1;
        }
    }

    // # of blocks in col0
    int col0 = std::popcount(state[0]);

    // bumpiness
    int bumpiness = 0;
    int diff = 0;

    for (int col = 0; col < width - 1; ++col)
    {
        diff = tops[col] - tops[col + 1];
        bumpiness += abs(diff);
    }

    // maxHeight, linesCleared

    std::array<int, 7> features = {-holes, -col0, -bumpiness, -maxHeight, -depth, linesCleared, tetris};

    double dot = 0;
    for (int i = 0; i < features.size(); ++i)
    {
        dot += (1.0 * features[i]) * weights[i];
    }
    return dot;
}

double Tetris::calculateNNScore()
{
    // # of holes
    int holes = 0;
    for (int col = 0; col < width; ++col)
    {
        holes += tops[col] - std::popcount(state[col]);
    }

    // depth of holes
    int depth = 0;
    int curr;
    int ones;
    for (int col = 0; col < width; ++col)
    {
        curr = state[col];
        ones = std::popcount(state[col]);3
        for (int row = 0; row < maxHeight; ++row)
        {
            if ((curr & 1) == 0)
            {
                depth += ones;
            }
            else
            {
                ones--;
            }
            curr >>= 1;
        }
    }

    // # of blocks in col0
    int col0 = std::popcount(state[0]);

    // bumpiness
    int bumpiness = 0;
    int diff = 0;

    for (int col = 0; col < width - 1; ++col)
    {
        diff = tops[col] - tops[col + 1];
        bumpiness += abs(diff);
    }

    // maxHeight, linesCleared

    std::array<int, 7> features = {-holes, -col0, -bumpiness, -maxHeight, -depth, linesCleared, tetris};

    // matmul1
    double intermediate1[5] = {0};
    for (int i = 0; i < 5; ++i)
    {
        for (int j = 0; j < 7; ++j)
        {
            intermediate1[i] += features[j] * w1[i][j];
        }
        intermediate1[i] += b1[i];
    }

    // RELU
    for (int i = 0; i < 5; ++i)
    {
        if (intermediate1[i] < 0)
            intermediate1[i] = 0;
    }

    // matmul2
    double intermediate2 = 0;
    for (int i = 0; i < 5; ++i)
    {
        intermediate2 += intermediate1[i] * w2[i];
    }

    intermediate2 += b2[0];

    // RELU
    if (intermediate2 < 0)
        intermediate2 = 0;

    return intermediate2;
}

void Tetris::updateSeed(int offset)
{
    seed += offset;
    rg.seed(seed);
}

bool Tetris::nextStateLookAhead()
{
    std::pair<int, int> bestMove = nextStateHelper(1);
    if (bestMove.first == -1 or totalLinesCleared >= 1000)
    {
        // DEAD
        return false;
    }
    else if (bestMove.first == -2)
    {
        // hold piece
        if (hold == -1)
        {
            hold = nextPiece;
            nextPiece = nextNextPiece;
            nextNextPiece = rg() % 7;
        }
        else
        {
            int temp = hold;
            hold = nextPiece;
            nextPiece = temp;
        }
        justHeld = true;
        return true;
    }
    // printState();
    drop(nextPiece, bestMove.first, bestMove.second);
    // printState();
    nextPiece = nextNextPiece;
    nextNextPiece = rg() % 7;
    justHeld = false;

    // prune
    if (igScore < -20)
    {
        std::cout << totalLinesCleared << "\t" << tetris << " pruned" << std::endl;
        return false;
    }

    tempLinesCleared += linesCleared;
    maxHeight = 0;

    if (tempLinesCleared > 5)
    {
        int rand = rg() % 10;
        int randLines = rg() % 4;
        for (int i = 0; i < width; ++i)
        {
            state[i] <<= randLines;
            if (i != rand)
            {
                state[i] |= ((1 << randLines) - 1);
            }
            tops[i] = 32 - std::countl_zero(state[i]);
            if (tops[i] > maxHeight)
                maxHeight = tops[i];
        }
        tempLinesCleared = 0;
    }
    return true;
}

std::pair<double, double> Tetris::nextStateHelper(int depth)
{
    if (depth == 1)
    {
        double nextScore;
        double bestScore = -__DBL_MAX__;
        std::pair<int, int> bestMove;
        Tetris nextBoard = *this;
        bool successful = false;
        for (int orientation = 0; orientation < allPieces[nextPiece].size(); ++orientation)
        {
            for (int xOffset = 0; xOffset < width - allPieces[nextPiece][orientation].width + 1; ++xOffset)
            {
                if (nextBoard.drop(nextPiece, orientation, xOffset))
                {
                    nextBoard.justHeld = false;
                    nextScore = nextBoard.nextStateHelper(2).first;
                    if (nextScore > bestScore)
                    {
                        bestScore = nextScore;
                        bestMove = {orientation, xOffset};
                    }
                    nextBoard = *this;
                    successful = true;
                }
            }
        }
        if (hold != -1 && justHeld == false)
        {
            for (int orientation = 0; orientation < allPieces[hold].size(); ++orientation)
            {
                for (int xOffset = 0; xOffset < width - allPieces[hold][orientation].width + 1; ++xOffset)
                {
                    if (nextBoard.drop(hold, orientation, xOffset))
                    {
                        // puts next piece to hold
                        nextBoard.hold = nextBoard.nextPiece;
                        nextBoard.justHeld = true;
                        nextScore = nextBoard.nextStateHelper(2).first;
                        if (nextScore > bestScore)
                        {
                            bestScore = nextScore;
                            bestMove = {-2, -2};
                        }
                        nextBoard = *this;
                        successful = true;
                    }
                }
            }
        }
        else if (justHeld == false)
        {
            // hold piece
            nextBoard.hold = nextBoard.nextPiece;
            nextBoard.justHeld = true;
            nextScore = nextBoard.nextStateHelper(2).first;
            if (nextScore > bestScore)
            {
                bestScore = nextScore;
                bestMove = {-2, -2};
            }
            nextBoard = *this;
            successful = true;
        }
        if (!successful)
        {
            return {-1, -1};
        }
        return bestMove;
    }
    else
    {
        double bestScore = -__DBL_MAX__;
        double nextNextScore;
        Tetris nextNextBoard = *this;
        for (int orientation = 0; orientation < allPieces[nextNextPiece].size(); ++orientation)
        {
            for (int xOffset = 0; xOffset < width - allPieces[nextNextPiece][orientation].width + 1; ++xOffset)
            {
                nextNextBoard.drop(nextNextPiece, orientation, xOffset);
                nextNextBoard.tetris += tetris;
                nextNextBoard.linesCleared += linesCleared;
                nextNextScore = nextNextBoard.calculateNNScore();
                if (nextNextScore > bestScore)
                {
                    bestScore = nextNextScore;
                }
                nextNextBoard = *this;
            }
        }
        // if there is a block in hold, drop held block
        if (hold != -1 && justHeld == false)
        {
            for (int orientation = 0; orientation < allPieces[hold].size(); ++orientation)
            {
                for (int xOffset = 0; xOffset < width - allPieces[hold][orientation].width + 1; ++xOffset)
                {
                    nextNextBoard.drop(hold, orientation, xOffset);
                    nextNextBoard.tetris += tetris;
                    nextNextBoard.linesCleared += linesCleared;
                    nextNextScore = nextNextBoard.calculateNNScore();
                    if (nextNextScore > bestScore)
                    {
                        bestScore = nextNextScore;
                    }
                    nextNextBoard = *this;
                }
            }
        }
        return {bestScore, 0};
    }
}

void Tetris::setPieces()
{
    Piece iPiece0;
    iPiece0.width = 4;
    iPiece0.skirt = {0, 0, 0, 0};
    iPiece0.blocks = {1, 1, 1, 1};

    Piece iPiece1;
    iPiece1.width = 1;
    iPiece1.skirt = {0, 0, 0, 0};
    iPiece1.blocks = {0b1111, 0, 0, 0};

    Piece jPiece0;
    jPiece0.width = 3;
    jPiece0.skirt = {0, 0, 0, 0};
    jPiece0.blocks = {0b11, 0b1, 0b1, 0};

    Piece jPiece1;
    jPiece1.width = 2;
    jPiece1.skirt = {0, 2, 0, 0};
    jPiece1.blocks = {0b111, 0b1, 0, 0};

    Piece jPiece2;
    jPiece2.width = 3;
    jPiece2.skirt = {1, 1, 0, 0};
    jPiece2.blocks = {0b1, 0b1, 0b11, 0};

    Piece jPiece3;
    jPiece3.width = 2;
    jPiece3.skirt = {0, 0, 0, 0};
    jPiece3.blocks = {1, 0b111, 0, 0};

    Piece lPiece0;
    lPiece0.width = 3;
    lPiece0.skirt = {0, 0, 0, 0};
    lPiece0.blocks = {1, 1, 0b11, 0};

    Piece lPiece1;
    lPiece1.width = 2;
    lPiece1.skirt = {0, 0, 0, 0};
    lPiece1.blocks = {0b111, 1, 0, 0};

    Piece lPiece2;
    lPiece2.width = 3;
    lPiece2.skirt = {0, 1, 1, 0};
    lPiece2.blocks = {0b11, 1, 1, 0};

    Piece lPiece3;
    lPiece3.width = 2;
    lPiece3.skirt = {2, 0, 0, 0};
    lPiece3.blocks = {1, 0b111, 0, 0};

    Piece oPiece0;
    oPiece0.width = 2;
    oPiece0.skirt = {0, 0, 0, 0};
    oPiece0.blocks = {0b11, 0b11, 0, 0};

    Piece sPiece0;
    sPiece0.width = 3;
    sPiece0.skirt = {0, 0, 1, 0};
    sPiece0.blocks = {1, 0b11, 1, 0};

    Piece sPiece1;
    sPiece1.width = 2;
    sPiece1.skirt = {1, 0, 0, 0};
    sPiece1.blocks = {0b11, 0b11, 0, 0};

    Piece tPiece0;
    tPiece0.width = 3;
    tPiece0.skirt = {0, 0, 0, 0};
    tPiece0.blocks = {1, 0b11, 1, 0};

    Piece tPiece1;
    tPiece1.width = 2;
    tPiece1.skirt = {0, 1, 0, 0};
    tPiece1.blocks = {0b111, 1, 0, 0};

    Piece tPiece2;
    tPiece2.width = 3;
    tPiece2.skirt = {1, 0, 1, 0};
    tPiece2.blocks = {1, 0b11, 1, 0};

    Piece tPiece3;
    tPiece3.width = 2;
    tPiece3.skirt = {1, 0, 0, 0};
    tPiece3.blocks = {1, 0b111, 0, 0};

    Piece zPiece0;
    zPiece0.width = 3;
    zPiece0.skirt = {1, 0, 0, 0};
    zPiece0.blocks = {1, 0b11, 1, 0};

    Piece zPiece1;
    zPiece1.width = 2;
    zPiece1.skirt = {0, 1, 0, 0};
    zPiece1.blocks = {0b11, 0b11, 0, 0};

    std::vector<Piece> iPieces = {iPiece0, iPiece1};
    std::vector<Piece> jPieces = {jPiece0, jPiece1, jPiece2, jPiece3};
    std::vector<Piece> lPieces = {lPiece0, lPiece1, lPiece2, lPiece3};
    std::vector<Piece> oPieces = {oPiece0};
    std::vector<Piece> sPieces = {sPiece0, sPiece1};
    std::vector<Piece> tPieces = {tPiece0, tPiece1, tPiece2, tPiece3};
    std::vector<Piece> zPieces = {zPiece0, zPiece1};

    allPieces = {iPieces, jPieces, lPieces, oPieces, sPieces, tPieces, zPieces};
}